lib/broadway/server.ex

```diff
@@ -153,10 +150,22 @@ defmodule Broadway.Server do
       end
 
     partitions = Keyword.keys(batchers_config)
-    dispatcher = {GenStage.PartitionDispatcher, partitions: partitions, hash: &{&1, &1.batcher}}
+
+    {type, dispatcher} =
+      case partitions do
+        [] ->
+          {:consumer, nil}
+
+        [_] ->
+          {:producer_consumer, GenStage.DemandDispatcher}
```

Comment:
@msaraiva I changed this to use pattern matching and only use partition dispatcher if partitions >= 2.